File: 2582-minimum-score-of-a-path-between-two-cities/minimum-score-of-a-path-between-two-cities.rb
# @param {Integer} n
# @param {Integer[][]} roads
# @return {Integer}
def min_score(n, roads)
  graph = Hash.new { |h, k| h[k] = [] }
  roads.each do |x,y,z|
    graph[x] << [y, z]
    graph[y] << [x, z]
  end
  queue = [1]
  vis = [false] * (n+1)
  mini = 10000000000
  while queue.size > 0
    node = queue.shift
    graph[node].each do |x,y|
      if vis[x] == false
        vis[x] = true 
        queue << x
      end
      mini = [mini, y].min
    end
  end
  mini
end

File: 2582-minimum-score-of-a-path-between-two-cities/minimum-score-of-a-path-between-two-cities.py
from collections import defaultdict, deque

class Solution:
    def minScore(self, n: int, roads: List[List[int]]) -> int:
        graph = defaultdict(list)

        for u, v, w in roads:
            graph[u].append((v, w))
            graph[v].append((u, w))

        visited = set()
        q = deque([1])
        ans = float('inf')

        while q:
            node = q.popleft()

            if node in visited:
                continue

            visited.add(node)

            for nei, weight in graph[node]:
                ans = min(ans, weight)   # consider edge in this component
                if nei not in visited:
                    q.append(nei)

        return ans